src/service/video-manager-impl.js

```diff
@@ -0,0 +1,318 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {listen, listenOnce, listenOncePromise} from '../event-helper';
+import {dev} from '../log';
+import {platformFor} from '../platform';
+import {fromClassForDoc} from '../service';
+import {VideoEvents, VideoAttributes} from '../video-interface';
+import {viewportFor} from '../viewport';
+import {vsyncFor} from '../vsync';
+
+/**
+ * @const {number} Percentage of the video that should be in viewport before it
+ * is considered visible.
+ */
+const VISIBILITY_PERCENT = 75;
+
+/**
+ * VideoManager keeps track of all AMP video players that implement
+ * the common Video API {@see ../video-interface.VideoInterface}.
+ *
+ * It is responsible for providing a unified user experience and analytics for
+ * all videos within a document.
+ */
+export class VideoManager {
+
+  /**
+   * @param {!./ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+
+    /** @private @const {!Window} */
+    this.win_ = ampdoc.win;
+
+    /** @private {?Array<!../video-interface.VideoInterface>} */
+    this.entries_ = null;
+
+    /** @private {boolean} */
+    this.scrollListenerInstalled_ = false;
+  }
+
+  /**
+   * Registers a video component that implements the VideoInterface.
+   * @param {!../video-interface.VideoInterface} video
+   */
+  register(video) {
+    dev().assert(video);
+
+    // TODO(aghassemi): Remove this later. For now, VideoManager only matters
+    // for autoplay videos so no point in registering arbitrary videos yet.
+    if (!video.element.hasAttribute(VideoAttributes.AUTOPLAY) ||
+        !platformSupportsAutoplay(platformFor(this.win_))) {
+      return;
+    }
+
+    if (!video.supportsPlatform()) {
+      return;
+    }
+
+    this.entries_ = this.entries_ || [];
```

Comment:
tiny allocation optimization since, at least for now, we only register autoplay videos, so many pages won't have anything to register hence the lazy initialization.